From 7d3e803e310c0d723e18e5b6b577383663432843 Mon Sep 17 00:00:00 2001
From: Philipp Schuster <philipp.schuster@cyberus-technology.de>
Date: Wed, 29 Nov 2023 15:48:51 +0100
Subject: [PATCH] grub-mkstandalone: warn on non-regular files

grub-mkstandalone silently dropped non-regular files, which could
cause unexpected omissions in standalone images. This commit:

- Updates documentation to clarify that graft point targets must be
  regular files.
- Emits a warning when a non-regular file is skipped.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

Index: 7d3e803e/util/grub-mkstandalone.c
===================================================================
--- 7d3e803e.orig/util/grub-mkstandalone.c
+++ 7d3e803e/util/grub-mkstandalone.c
@@ -115,7 +115,7 @@ argp_parser (int key, char *arg, struct
 
 struct argp argp = {
   options, argp_parser, N_("[OPTION] SOURCE..."),
-  N_("Generate a standalone image (containing all modules) in the selected format")"\v"N_("Graft point syntax (E.g. /boot/grub/grub.cfg=./grub.cfg) is accepted"),
+  N_("Generate a standalone image (containing all modules) in the selected format")"\v"N_("Graft point syntax (E.g. /boot/grub/grub.cfg=./grub.cfg) is accepted. The path on the right must point to a regular file."),
   NULL, help_filter, NULL
 };
 
@@ -191,8 +191,10 @@ add_tar_file (const char *from,
 
   COMPILE_TIME_ASSERT (sizeof (hd) == 512);
 
-  if (grub_util_is_special_file (from))
-    return;
+  if (grub_util_is_special_file (from)) {
+      grub_util_warn (_("skip: %s is not a regular file"), from);
+      return;
+  }
 
   optr = tcn = xmalloc (strlen (to) + 1);
   for (iptr = to; *iptr == '/'; iptr++);
